sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala

```diff
@@ -546,33 +281,33 @@ private[jdbc] class JDBCRDD(
       closed = true
     }
 
-    override def hasNext: Boolean = {
-      if (!finished) {
-        if (!gotNext) {
-          nextValue = getNext()
-          if (finished) {
-            close()
-          }
-          gotNext = true
-        }
-      }
-      !finished
-    }
+    context.addTaskCompletionListener{ context => close() }
```

Comment:
Note that `close()` is idempotent because it may be called twice, at the end of the task and at the end of iteration. In the old code the end-of-iteration case was handled in `hasNext`; here, it's handled in a `CompletionIterator` defined below.